### PR TITLE
SPQR- and BC-tree Improvements

### LINKED ISSUE #56: Virtual Edges in `DynamicPlanarSPQRTree` Skeletons are not initialized correctly
Consider the following MWE and its explanation below:
```cpp
#include <ogdf/decomposition/DynamicPlanarSPQRTree.h>
#include <ogdf/decomposition/StaticPlanarSPQRTree.h>

using namespace ogdf;

void printTree(SPQRTree &spqr) {
    const Graph &tree = spqr.tree();
    OGDF_ASSERT(tree.numberOfNodes() == 2);
    OGDF_ASSERT(spqr.typeOf(tree.nodes.head()) == SPQRTree::NodeType::SNode);
    OGDF_ASSERT(spqr.typeOf(tree.nodes.tail()) == SPQRTree::NodeType::RNode);

    Skeleton &sskel = spqr.skeleton(tree.nodes.head());
    Graph &skelG = sskel.getGraph();
    OGDF_ASSERT(skelG.numberOfNodes() == 3);
    OGDF_ASSERT(skelG.numberOfEdges() == 3);
    for (auto e : skelG.edges) {
        std::cout << "For root skeleton edge " << e << ":" << std::endl;
        edge real = sskel.realEdge(e);
        if (real != nullptr && reinterpret_cast<size_t>(real) < 0xFF)
            std::cout << "realEdge: WTF " << reinterpret_cast<size_t>(real) << ", " << std::endl;
        else
            std::cout << "realEdge: " << real << ", " << std::endl;
        std::cout << "isVirtual: " << sskel.isVirtual(e) << ", " << std::endl
                  << "twinEdge: " << sskel.twinEdge(e) << ", " << std::endl
                  << "twinTreeNode: " << sskel.twinTreeNode(e) << ", " << std::endl;
    }
}

int main() {
    Graph G;
    node nodes[]{
            G.newNode(),
            G.newNode(),
            G.newNode(),
            G.newNode(),
            G.newNode(),
    };
    edge edges[]{
            G.newEdge(nodes[0], nodes[1]),
            G.newEdge(nodes[1], nodes[3]),
            G.newEdge(nodes[2], nodes[0]),
            G.newEdge(nodes[3], nodes[2]),
            G.newEdge(nodes[2], nodes[4]),
            G.newEdge(nodes[0], nodes[3]),
            G.newEdge(nodes[4], nodes[1])
    };

//    {
//        std::cout << "Static:" << std::endl;
//        StaticPlanarSPQRTree sspqr(G);
//        printTree(sspqr);
//    }
    {
        std::cout << std::endl << "Dynamic:" << std::endl;
        DynamicPlanarSPQRTree dspqr(G);
        printTree(dspqr);
    }
}
```
It generates the SPQR tree for the following graph:
graph
The SPQR Tree looks like this:
spqr
It then iterates over the three edges in the skeleton of the S-node `S (0)` and prints whether they are virtual or not. Obviously, two of them should be real edges corresponding to the two edges in the upper left corner of the graph, while the third one is a virtual edge corresponding to the other R-node of the SPQR tree.
This is the output of the snippet on my system:
```
Dynamic:
For root skeleton edge (0,1):
realEdge: (2,4), 
isVirtual: 0, 
twinEdge: nil, 
twinTreeNode: nil, 
For root skeleton edge (1,2):
realEdge: (4,1), 
isVirtual: 0, 
twinEdge: nil, 
twinTreeNode: nil, 
For root skeleton edge (0,2):
realEdge: WTF 32, 
isVirtual: 0, 
twinEdge: (2,3), 
twinTreeNode: 1, 
```
The edge (0,2) is the virtual edge and correctly points to its twin node in the SPQR-tree. As a virtual edge, it should also have `realEdge=nullptr` and `isVirtual=1`, but the `realEdge` is actually a pointer to the low memory address `0x20` = 32 (while all the other edges reside in regions around `0x613f80`). Thus the value of `isVirtual` for the edge, which is calculated as `!realEdge(e)`, is also wrong. The function `realEdge` is defined [as follows](https://github.com/ogdf/ogdf/blob/master/src/ogdf/decomposition/DynamicSPQRTree.cpp#L59):
```cpp
edge DynamicSkeleton::realEdge(edge eM) const {
	return m_owner->m_hEdge_gEdge[m_origEdge[eM]];
}
```
It seems that the values in `m_hEdge_gEdge` are (unlike the values for twin edges in `m_hEdge_twinEdge`, which are set [here](https://github.com/ogdf/ogdf/blob/master/src/ogdf/decomposition/DynamicSPQRTree.cpp#L117) without touching the other array) not correctly initialized for virtual edges. As the array `m_hEdge_gEdge` is initialized without setting a default (see [here](https://github.com/ogdf/ogdf/blob/master/src/ogdf/decomposition/BCTree.cpp#L57)) and thus without overwriting any previous memory content, the `0x20` is probably some random left over memory value and this bug only appears randomly when the array is created in a memory region which is not filled with zeros beforehand. Changing the array initialization to `m_hEdge_gEdge.init(m_H, 0);` and rerunning the snippet again (after rebuilding the OGDF library) then yields the expected results:
```
Dynamic:
For root skeleton edge (0,1):
realEdge: (2,4), 
isVirtual: 0, 
twinEdge: nil, 
twinTreeNode: nil, 
For root skeleton edge (1,2):
realEdge: (4,1), 
isVirtual: 0, 
twinEdge: nil, 
twinTreeNode: nil, 
For root skeleton edge (0,2):
realEdge: nil, 
isVirtual: 1, 
twinEdge: (2,3), 
twinTreeNode: 1, 
```
If you don't want to add the overhead of filling the whole array with zeros, this bug could probably also be fixed by also setting the value in `m_hEdge_gEdge` to `nullptr` when [setting](https://github.com/ogdf/ogdf/blob/master/src/ogdf/decomposition/DynamicSPQRTree.cpp#L117) the `m_hEdge_twinEdge` for a virtual edge.

As side note: The code in the snippet yields correct results if you enable the lines with the static variant before the dynamic one, which also speaks for the "heisenbug"-like results of reading uninitialized memory. The outputs then are:
```
Static:
For root skeleton edge (0,1):
realEdge: (2,4), 
isVirtual: 0, 
twinEdge: nil, 
twinTreeNode: nil, 
For root skeleton edge (1,2):
realEdge: (4,1), 
isVirtual: 0, 
twinEdge: nil, 
twinTreeNode: nil, 
For root skeleton edge (2,0):
realEdge: nil, 
isVirtual: 1, 
twinEdge: (1,3), 
twinTreeNode: 1, 

Dynamic:
For root skeleton edge (0,1):
realEdge: (2,4), 
isVirtual: 0, 
twinEdge: nil, 
twinTreeNode: nil, 
For root skeleton edge (1,2):
realEdge: (4,1), 
isVirtual: 0, 
twinEdge: nil, 
twinTreeNode: nil, 
For root skeleton edge (0,2):
realEdge: nil, 
isVirtual: 1, 
twinEdge: (2,3), 
twinTreeNode: 1, 
```

### PR DESCRIPTION
**Pull request by N-Coder**
Title and commit messages say it all, nothing major here either. Might fix #56, but needs testing.

### CODE REVIEW

File: src/ogdf/decomposition/DynamicSPQRForest.cpp [hunk 1]
```diff
@@ -80,7 +80,7 @@ void DynamicSPQRForest::createSPQR(node vB) const {
 
 	Triconnectivity tricComp(GC);
 
-	const GraphCopySimple& GCC = *tricComp.m_pGC;
+	const GraphCopySimple& GCC = *dynamic_cast<GraphCopySimple*>(tricComp.m_pG);
```
Comment (milsen) [inline src/ogdf/decomposition/DynamicSPQRForest.cpp:83]: `dynamic_cast<const GraphCopySimple&>` should be possible afaik.

**Review by milsen — COMMENTED** [review]
Thank you for your work! :elephant:
Next to some suggestions, I'd also like that commit [clean-up Triconnectivity to make Graph usage consistent...](https://github.com/ogdf/ogdf/pull/94/commits/f67df19acba75c194cd4545faabedadcb07ac8c5) to have a shorter summary message. :D

File: include/ogdf/decomposition/DynamicSPQRForest.h [hunk 2]
```diff
@@ -327,19 +328,91 @@ class OGDF_EXPORT DynamicSPQRForest : public DynamicBCTree {
...
+	 * (e.g. using createSPQR(\p vB)) or 0 otherwise.
+	 */
+	int numberOfRNodes(node vB) const { return m_bNode_numR[vB]; }
+
+	//! Returns the SPQR-tree graph.
+	const Graph& spqrTree() const { return m_T; }
+
+	//! Returns the parent node of a node of the SPQR-tree Graph, or nullptr if \p vT is a root.
+	node spqrParent(node vT) const {
+		edge p_e = m_tNode_hRefEdge[vT];
```
[the remark this answers is not in the document]
Comment (N-Coder) [reply include/ogdf/decomposition/DynamicSPQRForest.h:398]: That should already be included in the operator[] of NodeArray

File: src/ogdf/graphalg/Triconnectivity.cpp [hunk 3]
```diff
@@ -61,62 +63,62 @@ Triconnectivity::Triconnectivity(const Graph& G) : m_ESTACK(G.numberOfEdges()) {
...
+			m_pG->reverseEdge(e);
 		}
 	}
 
 #ifdef OGDF_TRICONNECTIVITY_OUTPUT
 	std::cout << "\nnode\tNUMBER\tFATHER\tLOWPT1\tLOWPT2\tND" << std::endl;
-	for (node v : GC.nodes) {
-		std::cout << GC.original(v) << ":  \t" << m_NUMBER[v] << "   \t";
+	for (node v : m_pG->nodes) {
+		std::cout << GC().original(v) << ":  \t" << m_NUMBER[v] << "   \t";
```
[the remark this answers is not in the document]
Comment (N-Coder) [reply src/ogdf/graphalg/Triconnectivity.cpp:105]: True, but the zero-copy behaviour is also something that wasn't possible before and works with almost every method, except for the debug logging and checkComp. I did change the implementation as suggested and removed GC(). The logging will no no longer fail, but print the real nodes/edges. The problem with checkComp is that it relies on the information which edges where added by Triconnectivity, which is only implicitly tracked when using GraphCopy. I added in an assert to disable checkComp if this information is not available. All other uses should (still) be fine.

**Pull request merged**